src/codegen/compile.ml

```diff
@@ -9951,214 +9847,185 @@ let compile_binop env t op : SR.t * SR.t * G.t =
-                TaggedSmallWord.compile_nat_power env Type.Nat32 ^^ set_res ^^
+                let overflow_type = match ty with
+                | Type.Nat32 -> Type.Nat64 
+                | Type.(Nat8 | Nat16) -> Type.Nat32 
+                | _ -> assert false in
+                let overflow_type_bits = TaggedSmallWord.bits_of_type overflow_type in
+                let overflow_boundary = -Int.(sub (mul overflow_type_bits 2) 2) in
+                get_exp ^^ compile_unboxed_const 64L ^^
+                compile_comparison I64Op.GeU ^^ then_arithmetic_overflow env ^^
+                unsigned_dynamics get_n ^^ compile_sub_const (Int64.of_int bits) ^^
+                get_exp ^^ G.i (Binary (Wasm_exts.Values.I64 I64Op.Mul)) ^^
+                compile_unboxed_const (Int64.of_int overflow_boundary) ^^
+                compile_comparison I64Op.LtS ^^ then_arithmetic_overflow env ^^
+                get_n ^^ get_exp ^^
+                TaggedSmallWord.compile_nat_power env Type.Nat64 ^^ set_res ^^
                 get_res ^^ enforce_unsigned_bits env bits ^^
                 get_res ^^ TaggedSmallWord.msb_adjust ty
               end
-              get_n (* n@{0,1} ** (1+exp) == n *)
+              (get_n ^^ TaggedSmallWord.msb_adjust ty) (* n@{0,1} ** (1+exp) == n *)
           end
           (compile_unboxed_const
-             Int32.(shift_left one (to_int (TaggedSmallWord.shift_of_type ty))))) (* x ** 0 == 1 *)
-  | Type.(Prim Nat32),                        PowOp ->
-    Func.share_code2 Func.Always env (prim_fun_name Type.Nat32 "pow")
-      (("n", I32Type), ("exp", I32Type)) [I32Type]
-      (fun env get_n get_exp ->
-        let (set_res, get_res) = new_local64 env "res" in
-        get_exp ^^
-        G.if1 I32Type
-          begin
-            get_n ^^ compile_shrU_const 1l ^^
-            G.if1 I32Type
-              begin
-                get_exp ^^ compile_unboxed_const 32l ^^
-                G.i (Compare (Wasm.Values.I32 I32Op.GeU)) ^^ then_arithmetic_overflow env ^^
-                unsigned_dynamics get_n ^^ compile_sub_const 32l ^^
-                get_exp ^^ TaggedSmallWord.lsb_adjust Type.Nat32 ^^ G.i (Binary (Wasm.Values.I32 I32Op.Mul)) ^^
-                compile_unboxed_const (-62l) ^^
-                G.i (Compare (Wasm.Values.I32 I32Op.LtS)) ^^ then_arithmetic_overflow env ^^
-                get_n ^^ G.i (Convert (Wasm.Values.I64 I64Op.ExtendUI32)) ^^
-                get_exp ^^ G.i (Convert (Wasm.Values.I64 I64Op.ExtendUI32)) ^^
-                Word64.compile_unsigned_pow env ^^
-                set_res ^^ get_res ^^ enforce_32_unsigned_bits env ^^
-                get_res ^^ G.i (Convert (Wasm.Values.I32 I32Op.WrapI64))
-              end
-              get_n (* n@{0,1} ** (1+exp) == n *)
-          end
-          compile_unboxed_one) (* x ** 0 == 1 *)
-  | Type.(Prim ((Int8|Int16) as ty)),         PowOp ->
+             Int64.(shift_left one (to_int (TaggedSmallWord.shift_of_type ty))))) (* x ** 0 == 1 *)
```

Comment:
Simplify to
```
    (compile_unboxed_one ^^ TaggedSmallWord.msb_adjust ty) (* x ** 0 == 1 *)`
```

Reply:
Thanks! I applied the refactoring (in combination with the other occurrence above).